llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -9383,12 +9377,17 @@ void VPWidenLoadRecipe::execute(VPTransformState &State) {
   }
 }
 
+static Value *reverseMask(IRBuilderBase &Builder, Value *Mask, Value *EVL,
+                          Value *AllTrueMask) {
```

Comment:
>> Should the above reverse of the loaded values use the partial mask of these values, rather than an all-true mask?
>> 
>> Should the above reverse of the to-be-stored values use the partial mask of these values, rather than an all-true mask?
>
> Not necessary, all-true is good enough here, real mask just increases pressure on mask register with no effect

Worth a comment that the real mask is being optimized away _here_ by a cheaper AllTrueMask.

As noted, it's "Unclear if/when "masked reverse" should be used, when reversing masked operands", i.e., 
_where_ - for which reverses, if any? - all-true would not be good enough.

Would it be better to have `createReverseEVL(IRBuilderBase &Builder, Value *Operand, Value *EVL, const Twine &Name)` generate an `experimental_vp_reverse` for masks and values alike, also filling-in the optimized-out AllTrueMask?

Reply:
1.
> As noted, it's "Unclear if/when "masked reverse" should be used, when reversing masked operands", i.e., where - for which reverses, if any? - all-true would not be good enough.

The reversing of the values in registers is safe (the actual length of the reversed data is controlled by EVL, we just cut extra unused lanes, because of that we can safely use all-true mask). Does it answer your question?

2. Added createReverseEVL and used it.

Reply:
It's obviously safe for reverse to utilize an all-true mask - as does the non-EVL vectorizer. Question was if it is ever beneficial not to do so? I.e., if in practice the mask of reverse is an unused operand that only wastes a register.